cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -817,24 +818,158 @@ struct Log1pChecked {
+  static constexpr enable_if_floating_point<T> Round(T val) { return std::floor(val); }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_POS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(T val) { return std::ceil(val); }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(T val) { return std::trunc(val); }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(T val) {
+    return std::signbit(val) ? std::floor(val) : std::ceil(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_NEG_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(T val) {
+    return std::ceil(val - T(0.5));
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_POS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(T val) {
+    return std::floor(val + T(0.5));
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(T val) {
+    return std::copysign(std::ceil(std::fabs(val) - T(0.5)), val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_INFINITY> {
+  static enable_if_floating_point<T> Round(T val) {
+    return std::copysign(std::floor(std::fabs(val) + T(0.5)), val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_EVEN> {
+  static enable_if_floating_point<T> Round(T val) {
+    if (!RoundUtil::IsHalf(val)) {
+      return std::round(val);
+    }
+    auto floor = std::floor(val);
+    // Odd + 1, Even + 0
+    return floor + T(std::fmod(std::fabs(floor), T(2)) >= T(1));
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_ODD> {
+  static enable_if_floating_point<T> Round(T val) {
+    if (!RoundUtil::IsHalf(val)) {
+      return std::round(val);
+    }
+    auto floor = std::floor(val);
+    // Odd + 0, Even + 1
+    return floor + T(std::fmod(std::fabs(floor), T(2)) < T(1));
+  }
+};
+
+template <RoundMode RndMode>
+struct MRound {
+  template <typename T, typename Arg>
+  static enable_if_floating_point<Arg, T> Call(KernelContext* ctx, Arg arg, Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    if (std::isnan(arg)) {
+      return arg;
+    }
+    auto options = OptionsWrapper<MRoundOptions>::Get(ctx);
+    const auto mult = std::fabs(T(options.multiple));
+    return (mult == T(0)) ? T(0) : (RoundImpl<T, RndMode>::Round(arg / mult) * mult);
+  }
+};
+
+template <RoundMode RndMode>
+struct Round {
+  template <typename T, typename Arg>
+  static enable_if_floating_point<Arg, T> Call(KernelContext* ctx, Arg arg, Status*) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    auto options = OptionsWrapper<RoundOptions>::Get(ctx);
+    const auto mult = RoundUtil::Pow10<T>(-options.ndigits);
```

Comment:
We're not validating ndigits anywhere.

Reply:
`ndigits` and `multiple` can be any value (except NaN, Inf). If they are too large or too small they can produce `NaN` or `Inf` results. The "too large/small" will depend on the floating point type for output.

If there would be any validation, I recommend to do it when setting the `RoundOptions` value but given that these can be set directly to the data member (not through `set` methods), this is not feasible.

Reply:
ndigits gets used as an index in Pow10 so it needs validation - right now it'll just crash.

Reply:
Would you recommend resolving invalid values or explicitly triggering an error?
For cases where `ndigits` is out-of-range we can provide a corresponding meaningful result.

For example,
* If `ndigits` is "too large" and positive, which means to round to an extremely small decimal precision = `10^-(large number)` (most likely less than epsilon), then the rounded value is the input with no change.
* If `ndigits` is "too large" and negative, which means to round to an extremely large integral value = `10^(large number)` (most likely more than the range of floating-point values), then the rounded value is 0.

Reply:
I would say either a checked and unchecked variant, or else raise an error (since really, it's about the option being invalid). I don't think it's too useful to round to 0 or not round.

Reply:
Given that the `ndigits` value can be detected to be invalid in `Pow10(...)` helper function which is invoked by `Round::Call(...)`, I decided to propagate the error by returning a NaN value from `Pow10` and setting `Status::Invalid` if value is NaN. AFAIK, exceptions are not catched if not an "out-of-range" exception could be triggered via `std::array.at(...)`.

Reply:
We don't use exceptions; I think nan is fine.